### PR TITLE
UCT/IB/RC/DC: Set FAILED flag for TXWQ when error is detected

### PR DESCRIPTION
**Pull request by dmitrygx**
## What

Set flag `FAILED` for DCI when error is detected and verify that send/rma/amo/ep_check are not done when `FAILED` flag set on a DCI.

## Why ?

No operations should be scheduled on a failed DCI, because the operations will be failed immediately.

## How ?

1. Set `UCT_DC_MLX5_DCI_FLAG_FAILED` flag when failure is detected on some DCI.
2. Remove `UCT_DC_MLX5_DCI_FLAG_FAILED` flag in `uct_dc_mlx5_iface_reset_dci()`.
3. Check that `UCT_DC_MLX5_DCI_FLAG_FAILED` isn't set when doing am/rma/amo/ep_check operations.
4. Introduce `uct_dc_mlx5_iface_inline_post()` to minimize code duplication for inline operations.

### CODE REVIEW

**Comment by dmitrygx** [comment]
GPU failure was unrelated
@rakhmets @hoopoepg could you review pls?

**Review by hoopoepg — APPROVED** [review]

File: src/uct/ib/dc/dc_mlx5_ep.c [hunk 1]
```diff
@@ -308,13 +354,12 @@ ucs_status_t uct_dc_mlx5_ep_am_short(uct_ep_h tl_ep, uint8_t id, uint64_t hdr,
     uct_rc_mlx5_am_hdr_fill(&cache.am_hdr.rc_hdr, id);
     cache.am_hdr.am_hdr = hdr;
 
-    UCT_DC_MLX5_IFACE_TXQP_GET(iface, ep, txqp, txwq);
```
Comment (rakhmets) [inline src/uct/ib/dc/dc_mlx5_ep.c:311]: You can also remove `UCT_DC_MLX5_TXQP_DECL(txqp, txwq);` from the function.

**Review by rakhmets — APPROVED** [review]

File: src/uct/ib/dc/dc_mlx5_ep.c (hunk 1 above)
Comment (dmitrygx) [reply src/uct/ib/dc/dc_mlx5_ep.c:311]: thanks, good catch
fixed

**Comment by dmitrygx** [comment]
@yosefe could you review pls?

File: src/uct/ib/dc/dc_mlx5_ep.c [hunk 2]
```diff
@@ -19,6 +19,50 @@
...
+                              /* SEND */ uint8_t am_id, uint64_t hdr, uint32_t imm_val_be,
+                              /* RDMA */ uint64_t rdma_raddr, uct_rkey_t rdma_rkey,
+                              uint8_t send_flags, const void *buffer)
+{
+    UCT_DC_MLX5_TXQP_DECL(txqp, txwq);
+
+    ucs_assert(!(iface->tx.dcis[dci_index].flags &
+                         UCT_DC_MLX5_DCI_FLAG_FAILED));
+
+    UCT_DC_MLX5_IFACE_TXQP_DCI_GET(iface, dci_index, txqp, txwq);
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5_ep.c:35]: maybe we move the assert flag to txwq or txqp and the assert check into txqp_inline_post, then no need to add these wrappers, and will check all mlx5 transports?

Comment (dmitrygx) [reply src/uct/ib/dc/dc_mlx5_ep.c]: done

**Comment by dmitrygx** [comment]
we have UCT gtest which gets the EP failure and does send/flush on that EP:
https://github.com/openucx/ucx/blob/fd3e116e7f9ff61133918e1b06076d5d31f1ba6c/test/gtest/uct/test_peer_failure.cc#L338

what should we do with the test? is it correct behavior?

File: src/uct/ib/rc/base/rc_ep.c [hunk 3]
```diff
@@ -67,6 +71,32 @@ void uct_rc_txqp_cleanup(uct_rc_iface_t *iface, uct_rc_txqp_t *txqp)
...
+{
+#if UCS_ENABLE_ASSERT
+    txqp->flags = (txqp->flags | flags_add) & ~flags_remove;
+#endif
+}
+
+void uct_rc_txqp_post_flush_cancel_prologue(uct_rc_txqp_t *txqp,
+                                            unsigned flush_flags)
+{
+    if (ucs_unlikely(flush_flags & UCT_FLUSH_FLAG_CANCEL)) {
```
Comment (yosefe) [inline src/uct/ib/rc/base/rc_ep.c:85]: maybe just allow NOP opcode on failed QP?

File: src/uct/ib/rc/verbs/rc_verbs_ep.c [hunk 4]
```diff
@@ -37,6 +37,7 @@ uct_rc_verbs_ep_post_send(uct_rc_verbs_iface_t* iface, uct_rc_verbs_ep_t* ep,
 
     ucs_assertv(ep->qp->state == IBV_QPS_RTS, "QP 0x%x state is %d",
                 ep->qp->qp_num, ep->qp->state);
+    uct_rc_txqp_check_not_failed(&ep->super.txqp);
```
Comment (yosefe) [inline src/uct/ib/rc/verbs/rc_verbs_ep.c:40]: given the line above, we need the check only for mlx5 qps
so can move flags to txwq

File: src/uct/ib/rc/base/rc_ep.c (hunk 3 above)
Comment (dmitrygx) [reply src/uct/ib/rc/base/rc_ep.c]: done

File: src/uct/ib/rc/verbs/rc_verbs_ep.c (hunk 4 above)
Comment (dmitrygx) [reply src/uct/ib/rc/verbs/rc_verbs_ep.c]: thanks for catching this, done

File: test/apps/iodemo/io_demo.cc [hunk 5]
```diff
@@ -1316,6 +1316,39 @@ class DemoClient : public P2pDemoCommon {
         return data_size;
     }
 
+    void
+    get_server_info_uncompleted_ops(const server_info_t& server_info,
+                                    std::stringstream &ss) const
+    {
```
Comment (yosefe) [inline test/apps/iodemo/io_demo.cc:1322]: unrelated?

Comment (dmitrygx) [reply test/apps/iodemo/io_demo.cc]: yes, started re-working this PR on top of another PR by mistake.
fixed

File: src/uct/ib/mlx5/ib_mlx5.inl [hunk 6]
```diff
@@ -147,6 +147,26 @@ uct_ib_mlx5_txwq_validate(uct_ib_mlx5_txwq_t *wq, uint16_t num_bb)
...
+                              uint32_t flags_remove)
+{
+#if UCS_ENABLE_ASSERT
+    txwq->flags = (txwq->flags | flags_add) & ~flags_remove;
+#endif
+}
+
+
+static UCS_F_ALWAYS_INLINE void
+uct_ib_mlx5_txwq_check_not_failed(uct_ib_mlx5_txwq_t *txwq, uint8_t opcode)
```
Comment (yosefe) [inline src/uct/ib/mlx5/ib_mlx5.inl:161]: 1. no need this wrapper 
2. in common_post_send(): 
   if (opcode != NOP) { assert(...) }

**Review by yosefe — APPROVED** [review]

File: src/uct/ib/mlx5/ib_mlx5.inl (hunk 6 above)
Comment (dmitrygx) [reply src/uct/ib/mlx5/ib_mlx5.inl]: done

**Comment by dmitrygx** [comment]
https://github.com/openucx/ucx/pull/6928#issuecomment-859730670

@yosefe is it correct behavior? this test still fails.

**Comment by yosefe** [comment]
> [#6928 (comment)](https://github.com/openucx/ucx/pull/6928#issuecomment-859730670)
> 
> @yosefe is it correct behavior? this test still fails.

@dmitrygx this is a leftover from the days failed_ep was on UCT layer (so expected UCT-layer sending will failed), and not valid anymore now that failed_ep is on UCP layer

**Review by yosefe — APPROVED** [review]

File: src/uct/ib/rc/accel/rc_mlx5.inl [hunk 7]
```diff
@@ -450,6 +450,12 @@ uct_rc_mlx5_common_post_send(uct_rc_mlx5_iface_common_t *iface, int qp_type,
     struct mlx5_wqe_ctrl_seg *ctrl;
     uint16_t res_count;
 
+    if (opcode == MLX5_OPCODE_NOP) {
```
Comment (yosefe) [inline src/uct/ib/rc/accel/rc_mlx5.inl:453]: it should be != NOP

**Review by yosefe — CHANGES_REQUESTED** [review]

File: src/uct/ib/rc/accel/rc_mlx5.inl (hunk 7 above)
Comment (dmitrygx) [reply src/uct/ib/rc/accel/rc_mlx5.inl]: fixed

**Review by yosefe — APPROVED** [review]

**Comment by dmitrygx** [comment]
n

> > [#6928 (comment)](https://github.com/openucx/ucx/pull/6928#issuecomment-859730670)
> > @yosefe is it correct behavior? this test still fails.
> 
> @dmitrygx this is a leftover from the days failed_ep was on UCT layer (so expected UCT-layer sending will fail), and not valid anymore now that failed_ep is on UCP layer

I see, shall we remove the test then as it is not valid anymore?

**Comment by yosefe** [comment]
> I see, shall we remove the test then as it is not valid anymore?

Yes

**Comment by dmitrygx** [comment]
> > I see, shall we remove the test then as it is not valid anymore?
> 
> Yes

done + squashed

**Review by yosefe — APPROVED** [review]

**Comment by dmitrygx** [comment]
@yosefe it was necessary to fix one more test by doing `ep_check` instead of `ep_am_short`
could you review the 2nd commit, please?

**Comment by yosefe** [comment]
@dmitrygx what was the problem with send_am? maybe the test should avoid doing it for ep which already failed?

**Comment by dmitrygx** [comment]
@yosefe could you review pls the `test/great/cut/test_peer_failure.cc` and `test/great/cut/test_peer_failure.h` files? I’ve added checking failed EPs and do send_am for not failed ones only.
Thanks!

File: test/gtest/uct/test_peer_failure.cc [hunk 8]
```diff
@@ -140,10 +141,17 @@ void test_uct_peer_failure::set_am_handlers()
...
+
     while ((status = uct_ep_am_short(m_sender->ep(index), 0, 0, NULL, 0)) ==
            UCS_ERR_NO_RESOURCE) {
         progress();
-    };
+
+        status = ep_status(index);
+        if (status != UCS_OK) {
+            break;
+        }
```
Comment (yosefe) [inline test/gtest/uct/test_peer_failure.cc:152]: why needed since we are not supposed to call send_am on failed ep?

Comment (dmitrygx) [reply test/gtest/uct/test_peer_failure.cc:152]: We do progress(), so the failure on EP could be detected after progress

Comment (yosefe) [reply test/gtest/uct/test_peer_failure.cc:152]: in which line?

Comment (dmitrygx) [reply test/gtest/uct/test_peer_failure.cc:152]: we do progress() on `test/gtest/uct/test_peer_failure.cc:147`

Comment (yosefe) [reply test/gtest/uct/test_peer_failure.cc:152]: why is it important to return ep_status from this function, and can't just return the result of uct_ep_am_short?

Comment (dmitrygx) [reply test/gtest/uct/test_peer_failure.cc:152]: uct_ep_am_short will return UCS_OK, even if error detected

Comment (yosefe) [reply test/gtest/uct/test_peer_failure.cc:152]: but aren't we skipping send_am if error is detected?

Comment (dmitrygx) [reply test/gtest/uct/test_peer_failure.cc:152]: > but aren't we skipping send_am if error is detected?

yes, we do it
but the error can be detected during doing progress in `send_am`

Comment (yosefe) [reply test/gtest/uct/test_peer_failure.cc:152]: so maybe move checking failed eps set logic to send_am while loop (and remove ep_status wrapper), and also remove check before calling send_am?

Comment (dmitrygx) [reply test/gtest/uct/test_peer_failure.cc:152]: > so maybe move checking failed eps set logic to send_am while loop (and remove ep_status wrapper), and also remove check before calling send_am?

but we should if the EP failed before doing uct_ep_am_short.
also, we should return the status instead of `UCS_OK` from `send_am` to stop doing futher `send_am`.

wdyt about the following?
```
do {
        progress();

        status = ep_status(index);
        if (status != UCS_OK) {
            break;
         }
} while (uct_ep_am_short(m_sender->ep(index), 0, 0, NULL, 0) == UCS_ERR_NO_RESOURCE);

return status;
```

Comment (yosefe) [reply test/gtest/uct/test_peer_failure.cc:152]: (suggestion only)
```
for (;;) {
   /* if endpoint has failed, return error and avoid calling send
   it = find(..)
   if (!= end) 
     return it->second
   status = am_send()
  progress
}
```

Comment (dmitrygx) [reply test/gtest/uct/test_peer_failure.cc]: done

**Review by yosefe — APPROVED** [review]
